# SCIPhI: Single-cell mutation identification via phylogenetic inference
#
# Copyright (C) 2018 ETH Zurich, Jochen Singer
#
# This file is part of SCIPhI.
#
# SCIPhI is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# SCIPhI is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with SCIPhI. If not, see <http://www.gnu.org/licenses/>.
# 
# @author: Jochen Singer

import os, glob, sys
from snakemake.utils import R

SAMPLEMAPPING = 'sample_mapping.txt' 
OUTDIR   = 'out/'

ORGANISM = 'H_sapiens_hg19'

FASTQDIR = ""
BWAIN = ""
BWAOUT = ""
FIXMATEANDSORTIN = ""
FIXMATEANDSORTOUT = ""
MERGEBAMSOUT = ""
REMOVEPCRDUBLICATESOUT = ""
REALIGNINDELSOUT = OUTDIR + 'realignedIndels/'
MPILEUPIN = REALIGNINDELSOUT
MPILEUPOUT = OUTDIR + 'mpileup/'
SCIPHIOUT = OUTDIR + 'variants/sciphi/'
MONOVAROUT = OUTDIR + 'variants/monovar/'
FINALBAMOUTDIR = REALIGNINDELSOUT

include: "../../pipe/misc_snake.py"

TUMORSAMPLENAMES = getTumorSampleNames("CT")
CELLNORMALSAMPLENAMES = getTumorSampleNames("CN")
BULKNORMALSAMPLENAMES = getTumorSampleNames("BN")
SAMPLENAMES = CELLNORMALSAMPLENAMES + BULKNORMALSAMPLENAMES + TUMORSAMPLENAMES
PAIREDFASTQFILESWITHOUTR = getPairedFastqFilesWithoutR(SAMPLENAMES)
SINGLEFASTQFILES = ""

EXP = 'patient9ovarian'

include: "../../pipe/align_snake.py"
include: "../../pipe/monovar_snake.py"
include: "../ovarian.snake"

localrules: all
rule all:
    input:
        expand(OUTDIR + 'variants/monovar/{experiment}.vcf', experiment = getExperimentNames()),
        expand(OUTDIR + 'variants/sciphi/{run}/'+ EXP + '.vcf', run = ['1', '2', '3']),
        expand(OUTDIR + 'variants/sciphi/{run}/{experiment}_heatTree.pdf', experiment = getExperimentNames(), run = ['1', '2', '3']),
        expand(OUTDIR + 'variants/monovar/{experiment}_heatTree.pdf', experiment = getExperimentNames()),
        expand(OUTDIR + 'variants/monovar/{experiment}_cluster.pdf', experiment = getExperimentNames())

